gadopt/stokes_integrators.py

```diff
@@ -193,6 +211,62 @@ def setup_solver(self):
         self._solver_setup = True
 
     def solve(self):
+        if not self._initialised:
+            self.initialise()
         if not self._solver_setup:
             self.setup_solver()
+
+        self.solution_old.assign(self.solution)  # Need to update old solution for implicit free surface
+
         self.solver.solve()
+
+
+class FreeSurfaceStokesSolver(StokesSolver):
+    name = 'FreeSurfaceStokesSolver'
+
+    def __init__(self, z, T, approximation, free_surface_dt, free_surface_id, bcs=None, mu=1,
+                 quad_degree=6, cartesian=True, solver_parameters=None, closed=True,
+                 rotational=False, J=None, constant_jacobian=False, **kwargs):
+
+        super().__init__(z, T, approximation, bcs=bcs, mu=mu, quad_degree=quad_degree,
+                         cartesian=cartesian, solver_parameters=solver_parameters, closed=closed,
+                         rotational=rotational, J=J, constant_jacobian=constant_jacobian, **kwargs)
+
+        self.free_surface_dt = free_surface_dt
+        self.free_surface_id = free_surface_id
+
+    def initialise(self):
+        # Solution is made up of 3 trial functions so we have a separate initialise()
+        u, p, eta = fd.split(self.solution)
+        u_old, p_old, eta_old = fd.split(self.solution_old)
+
+        # Theta timestepping scheme for free surface variable
+        # Forward Euler (theta = 0), Crank Nicholson (thetat = 0.5), Backward Euler (theta = 1)
+        theta = 0.5
+        eta_theta = (1-theta)*eta_old + theta*eta
```

Comment:
So looking at how you use eta, and how you do time loops, I wonder if there is a way to be consistent with how you are doing your time integration. Here you are hard coding `$\theta=0.5$`, but then when time-stepping in your actual time integration you often use `BackwardEuler`. 

Correct me if I am wrong, but the correct way of doing this would be to supply a time-stepping scheme to `FreeSurfaceStokesSolver` that will self-consistently first solve Stokes with the consistent approximation of `eta` in time, and then advance `eta` with the same approximation.

Reply:
I think we discussed this separately yesterday so hopefully it is ok.

One thing is whether the last equation should be multiplied by theta so that the u, p, eta block is symmetric... 

e.g from Stephan's 2012 paper:

![image](https://github.com/g-adopt/g-adopt/assets/23333139/88e8a467-0600-4090-bfb0-7b15eea964b3)

Reply:
So I have updated the code now so that the bottom equation is multiplied by theta but i think I shouldnt have multiplied the u.n term by delta t...

Reply:
So i think now it is better with only theta multiplied by u.n... but should is it now antisymmetric? do i need a minus sign...

Reply:
@w3168 Just checking if this conversation should be resolved.

Reply:
@Patol75  not yet! though stephan brought it up again so I think it should be resolved soon!